drivers/net/ncv7410.c

```diff
@@ -0,0 +1,1669 @@
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.  The
+ * ASF licenses this file to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance with the
+ * License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.  See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ *
+ ****************************************************************************/
+
+/****************************************************************************
+ * Included Files
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <stdint.h>
+#include <stdbool.h>
+#include <string.h>
+#include <assert.h>
+#include <debug.h>
+#include <errno.h>
+
+#include <nuttx/spi/spi.h>
+#include <sys/endian.h>
+
+#include <nuttx/irq.h>
+#include <nuttx/kmalloc.h>
+#include <nuttx/wqueue.h>
+#include <nuttx/signal.h>
+#include <nuttx/mutex.h>
+#include <nuttx/net/ncv7410.h>
+#include <nuttx/net/netdev_lowerhalf.h>
+
+#include "ncv7410.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
+ ****************************************************************************/
+
+#define NCVWORK LPWORK
+
+#define NCV_RESET_TRIES 5
+
+/* Maximum frame size = (MTU + LL heaader size) + FCS size */
+
+#define NCV_MAX_FRAME_SIZE(p) (p->dev.netdev.d_pktsize + 4)
+
+/* Packet Memory ************************************************************/
+
+/* Maximum number of allocated tx and rx packets */
+
+#define NCV7410_TX_QUOTA        1
+#define NCV7410_RX_QUOTA        2
+
+#if CONFIG_IOB_NBUFFERS < (NCV7410_TX_QUOTA + NCV7410_RX_QUOTA)
+#  error "CONFIG_IOB_NBUFFERS must be > (NCV7410_TX_QUOTA + NCV7410_RX_QUOTA)"
+#endif
+
+#ifndef CONFIG_SCHED_LPWORK
+#  error "CONFIG_SCHED_LPWORK is needed by NCV7410 driver"
+#endif
+
+/****************************************************************************
+ * Private Types
+ ****************************************************************************/
+
+/* The ncv7410_driver_s encapsulates all state information for a single
+ * hardware interface
+ */
+
+enum ncv_ifstate_e
+{
+  NCV_RESET,
+  NCV_INIT_DOWN,
+  NCV_INIT_UP
+};
+
+struct ncv7410_driver_s
+{
+  struct netdev_lowerhalf_s dev; /* Driver data visible by the net stack
+                                  * (must be placed first)                 */
+  mutex_t mutex;                 /* Mutex for data race prevention         */
```

Comment:
@matiamic just a minor modification before we merge it: could you please rename to: "mutex_t lock;" ?
If you look inside NuttX code this is the preferred form, in the past the lock was a semaphore, and we don't know if it could change in the future. Calling it "lock" make it independent of the used solution.